### PR TITLE
KAFKA-14365: Refactor Fetcher to allow different implementations

### PR DESCRIPTION
**Pull request by kirktrue**
This change refactors the existing `Fetcher` by splitting out the parts related to offset listing into a dedicated class named `OffsetsFinder`. This allows the `Fetcher` implementation to be more dedicated and thus easier to refactor (in coming commits).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+    }
+
+    /**
+     * Get metadata for all topics present in Kafka cluster
+     *
+     * @param request The MetadataRequest to send
+     * @param timer Timer bounding how long this method can block
+     * @return The map of topics with their partition information
+     */
+    public Map<String, List<PartitionInfo>> getTopicMetadata(MetadataRequest.Builder request, Timer timer) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: This doesn't seem related to finding offsets. Should it be somewhere else?

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: Agreed. Is adding it to `KafkaConsumer` directly OK?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+import java.util.stream.Collectors;
+
+public class OffsetsFinder {
+    private final Logger log;
+    private final ConsumerNetworkClient client;
+    private final Time time;
+    private final long retryBackoffMs;
+    private final long requestTimeoutMs;
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptions;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: To get a real benefit out of this refactor, we need to reduce the dependence on some of these components. Rather than pulling state directly from `SubscriptionState`, for example, it would be better to have methods which accept the state we need as arguments. Then we can focus this component on the functional logic, such as building requests, checking errors, and handling backoff. We also, want to drop the dependence on `ConsumerNetworkClient`. All the sending/polling will be done in the background thread.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: We can generalize the name perhaps? Instead of `OffsetFinder`, perhaps we can call it `MetadataManager` or something generic like that.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java (hunk 2 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: Agreed!

Splitting the functionality was introduced as a stand-alone PR in an attempt to have more and smaller PRs vs. fewer and larger PRs. This was to be just the first PR of several.

Would you prefer to have everything fleshed out completely?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java (hunk 1 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java]: Name has been changed on the latest commit.
